feat(DatePicker): add quick selection

tomru opened the pull request:
## Adds quick selection ranges to the Date Picker.

The DatePicker can now offer _quick selections_ for ranges to the user.

These quick selections can be specified by setting the prop `rangeQuickSelections` in the form of e.g. 
```js
[
  {
    label: 'Today',
    range: 'P1D',
  },
  {
    label: '7D',
    range: 'P7D',
  },
  {
    label: '14D',
    range: 'P14D',
  },
  {
    label: '1M',
    range: 'P1M',
  },
  {
    label: '2M',
    range: 'P2M',
  },
]
```

The quick selection ranges are only shown if the component has
* `rangeSelect`
* `rangeQuickSelections` not set to `null` or an empty array.

When the user clicks on of these the `onSelect` callback is called with the `range` of the quick selection clicked.  It's then up to the application to convert the `range` into `startDate` and `endDate` and update these properties on the `DatePicker` so that the clicked range is selected.

This allows us to keep the application logic outside of axiom and keep the implementation completely agnostic of what these ranges mean in an application. (i.e. does 7D means 7*24h ago until now or start of the day 7 days ago until now....)

![single](https://user-images.githubusercontent.com/111471/44025304-088c9110-9ef0-11e8-8ed7-8e59979ee50d.gif)

![double](https://user-images.githubusercontent.com/111471/44025346-2d3b712a-9ef0-11e8-9a96-3eceb1783941.gif)


~~[Netlify Link](https://5b71551b4ed62f6d0c396c7e--youthful-albattani-41ebb2.netlify.com/docs/packages/axiom-components/date-picker)~~
~~[Netlify Link](https://5b72854dc96592106b141531--youthful-albattani-41ebb2.netlify.com/docs/packages/axiom-components/date-picker)~~
~~[Netlify Link](https://5b728f081f12b744b73c03d0--youthful-albattani-41ebb2.netlify.com/docs/packages/axiom-components/date-picker/)~~
[Netlify Link](https://5b87e5191f12b70f1a6bfb06--brandwatch-axiom-tomru.netlify.com/docs/packages/axiom-components/date-picker/)

tomru commented:
> Note: The quick selection Today does not set a dateEnd just a dateStart analog to the behaviour when you just click a single day.

Just noticing that you can set a range of the same day, when you click twice on the day. So 'TODAY' could do the same.

tomru commented on packages/axiom-components/src/DatePicker/DatePicker.js:
```diff
@@ -39,6 +45,28 @@ export default class DatePicker extends Component {
+    rangeQuickSelections: [
+      {
+        label: 'Today',
+        daysBeforeNow: 0,
```
I would have loved to go with [ISO 8601 durations](https://en.wikipedia.org/wiki/ISO_8601#Durations) here but date-fns does not yet support these :(

HHogg replied:
@stefanpearson recommended us switching over to [luxon](https://github.com/moment/luxon) which does have support for it. Maybe now's a good time?

tomru replied:
Good call, will have a look into that 👍

HHogg commented on packages/axiom-components/src/DatePicker/DatePicker.js:
```diff
@@ -24,6 +24,12 @@ export default class DatePicker extends Component {
+    rangeQuickSelections: PropTypes.arrayOf(PropTypes.shape({
+      label: PropTypes.string.isRequired,
+      daysBeforeNow: PropTypes.number,
+      monthsBeforeNow: PropTypes.number,
```
I feel like switching over to something that supports the ISO 8601 durations would also make these props a bit nicer, as they could be a bit more generic and could support future ranges, years, hours, minutes etc...

HHogg commented on packages/axiom-components/src/DatePicker/DatePickerQuickSelection.js:
```diff
@@ -0,0 +1,38 @@
+  };
+
+  constructor(props) {
+    super(props);
```
You don't need this constructor here :)

HHogg commented on packages/axiom-components/src/DatePicker/DatePickerQuickSelection.css:
```diff
@@ -0,0 +1,41 @@
+  }
+
+  &:hover {
+    color: var(--color-ui-accent);
```
It would be nice to have some transitions here between the colours... ie 

```css
transition: color var(--transition-time-fast) var(--transition-function);
```

HHogg commented:
I think the double view is removed below a certain breakpoint.... maybe we should look into how to get the range selections to work on the single view. Ping @mozzling @Flawwles 

<img width="436" alt="screen shot 2018-07-05 at 14 45 18" src="https://user-images.githubusercontent.com/7130591/42326948-7ba79192-8062-11e8-92f6-23417fcb41c6.png">

tomru commented:
@Flawwles struggling a little bit with this as the behaviour of quick selection is unclear to me

The design assumes the quick selection is a short cut for selecting date ranges manually, hence the selected date range should be shown as if you would have selected them manually. But this raises the question what these ranges actually mean, e.g. 
- is `Today`, start of today to now or start of today to end of today
- is `7D`,  today and the last 6 full days, or as in analytics today + the last 7 full days or even exactly 7 days ago until now (as far as I can tell vizia does that different currently as well @TomWija right?)

This is IMO this is down to the application logic what these quick selection mean. Hence we should not define in Axiom what these ranges are.

On top we can just visually select the day the range starts and the day the range ends. The user has no clue yet, what time on that day the range actually starts.

To get around defining what Today, 7D, etc. means can we make the quick selection rather a first class selection? Meaning when you select 7D, we won't highlight what 7D means in the picker?

In that case I think the design needs to emphasise a bit more that we have selected a range instead of a start and end date. Maybe show a textual description, like `Last 7 days` where the start and end date would be shown when you select them manually?

tomru commented:
@Flawwles You can ignore my last comment, as I've now changed the implementation to be free of deciding what e.g. 7D means. This is now up to the application consuming the DatePicker. Can you please review this on the netlify link in the description please?

@HHogg I've refactored this quite a bit now to keep the application logic out of axiom. This also means that in the DatePicker we do not actually use luxon to interpret the ISO 8061 durations anymore.

tomru commented on site/components/Documentation/Resources/Components/DatePicker.js:
```diff
@@ -28,12 +68,16 @@ export default class Documentation extends Component {
+                    this.handleSelect({ date, dateEnd, dateStart, range });
+                  } }
+                  rangeSelect
+                  rangeSelections={ rangeSelections }
```
I've added the `rangeSelections` to both examples. 

As far as I can see on example is showing single view the other double view. Wondering if I should rather make 4 examples covering all variants, or rather drop one and rely on _you need to modify the settings_.

I think to show how to use it, it would be good add the `rangeSelections` at least at one place though...

tomru commented:
Now rebased after https://github.com/BrandwatchLtd/axiom/pull/628 was merged.

lpoulter commented on packages/axiom-components/src/DatePicker/DatePickerRangeSelection.css:
```diff
@@ -0,0 +1,45 @@
+.ax-date-range-selection {
```
There's a bit of jumping around here when the month spans 5 or 6 rows.

tomru replied:
True, I've noticed that as well. Though this hasn't really change from how it was before as far as I can tell.

Even when we would reserve the space for 6 rows to get rid of the jumping, it would look weird in the 4 col case (which applies more often).

Any thoughts on this @Flawwles ?

lpoulter commented on packages/axiom-components/src/DatePicker/DatePicker.js:
```diff
@@ -22,14 +22,25 @@ export default class DatePicker extends Component {
+      }),
+    ),
     /** A single date that appears selected */
     selectedDate: PropTypes.instanceOf(Date),
```
Getting some prop type errors on local. `selectedDate` is being set to a blank string when in range mode I think. Which is then passed into the child components.

lpoulter commented on packages/axiom-components/src/DatePicker/DatePickerContext.js:
```diff
@@ -59,6 +68,22 @@ export default class DatePickerContext extends Component {
 
+  componentDidUpdate(prevProps) {
+    const { selectedStartDate, selectedEndDate, view } = this.props;
+
```
Not sure I really understand why we need to mirror the props `selectedStartDate` & `selectedEndDate` into their `active` equivalents state here?

tomru replied:
The `active[Start|End]Date` do control which month is shown on both sides of the picker.

When setting `selected[Start|End]Date` props on the picker, after selecting a month, the user should see the start and end of the selected range. Hence I need to update both moth grids as well.

lpoulter commented on packages/axiom-components/src/DatePicker/DatePickerContext.js:
```diff
@@ -127,7 +167,46 @@ export default class DatePickerContext extends Component {
+    const isDoubleView = view === 'double';
+    const showRangeSelection = rangeSelect && rangeSelections && rangeSelections.length;
+
+    const rangeSelection = showRangeSelection && (isDoubleView ? (
```
Theres a lot of logic the pivots on weather `rangeSelect` is applied. Have you tried splitting out some of this into a separate component? The switching between a single date, `selectedDate` and a range of dates, `selectedStartDate` & `selectedEndDate` is quite hard to follow.

lpoulter commented on packages/axiom-components/src/DatePicker/DatePickerContext.js:
```diff
@@ -105,6 +130,19 @@ export default class DatePickerContext extends Component {
+      rangeSelect,
+    } = this.props;
+
+    if (!rangeSelect || !range) {
```
Look like we are already checking this in the render.

tomru replied:
Depends, if you pass `rangeSelections` according to the prop types. But in that case it's probably your fault as well, when `onSelect` is called without a range value. 

Dropping 👍

lpoulter commented on packages/axiom-components/src/DatePicker/DatePickerContext.js:
```diff
@@ -127,7 +167,46 @@ export default class DatePickerContext extends Component {
+    const isDoubleView = view === 'double';
+    const showRangeSelection = rangeSelect && rangeSelections && rangeSelections.length;
+
+    const rangeSelection = showRangeSelection && (isDoubleView ? (
```
`showRangeSelection && (isDoubleView ? (` is the `(` before isDoubleView doing any thing here?

lpoulter replied:
Perhaps this template could be simplified if we had a `DatePickerView` component that rendered one or two DatePickerHeaderControl/DatePickerViewMonth pairs based on a double prop.

tomru replied:
`&&` has a higher operator precedence than the conditional operator, so I'm forced to use a bracket here if I'm not missing anything.

tomru replied:
Might be worth splitting it out, but I'm concerned it won't reduce the complexity a lot, it would just move it to another file. I think to reduce the complexity it might be worth to to split the date picker into a `SingleDatePicker` and a `RangeDatePicker` to reduce the different cases.

The `SingleDatePicker` could be free of all the range related stuff, as well as the `RangeDatePicker` could be free of the single date stuff ?

lpoulter replied:
I've not seen this sort of ternary before. This is confusing but I think there is a bug here.
```
let showRangeSelection = false;
let isDoubleView = false;
showRangeSelection && (isDoubleView ? 'trueCase' : 'falseCase');
returns false
```

lpoulter replied:
Either way I think it would be clear if it was split out.

tomru replied:
doh! brackets belong here `(showRangeSelection && isDoubleView) ? ... ` 

Thanks for pointing that out!

tomru commented on packages/axiom-components/src/DatePicker/DatePickerContext.js:
```diff
@@ -59,6 +68,22 @@ export default class DatePickerContext extends Component {
+      this.setState({ activeStartDate: selectedStartDate });
+    }
+
+    if (view === 'double' && selectedEndDate !== prevProps.selectedEndDate) {
```
As this part only sets `activeEndDate` which affects only the second month grid, I think it makes sense to check if we are in double view. I agree it makes sense to check for `rangeSelect` additionally to keep the state clean.

tomru commented on packages/axiom-components/src/DatePicker/DatePickerContext.js:
```diff
@@ -127,7 +167,46 @@ export default class DatePickerContext extends Component {
+    const isDoubleView = view === 'double';
+    const showRangeSelection = rangeSelect && rangeSelections && rangeSelections.length;
+
+    const rangeSelection = showRangeSelection && (isDoubleView ? (
```
Yeah, it's not pretty, but couldn't think of another way of doing this currently.

tomru commented:
@lpoulter @HHogg I've split up `DatePickerContext` a bit, by
- moving the gnarly conditional for responsiveness down into `DatePickerRangeSelection`
- Extracted displaying single and double _day in month pickers_ into `DatePickerSelection` - I'm still unsure about the the name tbh

Could you have another look please ?

tomru commented:
@Flawwles IIRC the thing left was the question if the documentation is ok ( one opinion [here](https://github.com/BrandwatchLtd/axiom/pull/614#discussion_r209947160) ). Anything else that needs to be addressed ?

HHogg commented on packages/axiom-components/src/DatePicker/SelectedDateOrRangePropTypes.js:
```diff
@@ -0,0 +1,44 @@
+    return;
+  }
+
+  if (value instanceof Date && !isNaN(value)) {
```
Could these custom propTypes make use `PropTypes.instanceOf` somehow?

tomru replied:
They could, but it wouldn't detect
- invalid dates
- invalid combinations of `rangeSelect`, `selectedDate`, `selectedStartDate` and `selectedEndDate`

So I think this should be superiorly
